### PR TITLE
Refactor HSM task interface to support new use cases

### PR DESCRIPTION
**Pull request by bergundy**
## What changed?

Removed the `Kind` concept and allow tasks to be both outbound timer and immediate tasks.

## Why?

Simplification, prep for future capabilities.

## How did you test it?

Existing tests.

### CODE REVIEW

File: service/history/hsm/tasks.go [hunk 1]
```diff
@@ -27,6 +27,8 @@ import (
 	"time"
 )
 
+var NoDeadline = time.Time{}
```
Comment (bergundy) [inline service/history/hsm/tasks.go:30]: I should probably rename this to `Immediate`.

File: service/history/hsm/tasks.go [hunk 2]
```diff
@@ -27,6 +27,8 @@ import (
...
+
 // ErrInvalidTaskKind can be returned by a [TaskSerializer] if it received the wrong task kind.
 var ErrInvalidTaskKind = errors.New("invalid task kind")
```
Comment (bergundy) [inline service/history/hsm/tasks.go:33]: And this can be removed.

File: components/callbacks/statemachine.go [hunk 3]
```diff
@@ -94,21 +94,22 @@ func (c Callback) recordAttempt(ts time.Time) {
...
+		var destination string
 		switch v := c.Callback.GetVariant().(type) {
 		case *persistencespb.Callback_Nexus_:
 			u, err := url.Parse(c.Callback.GetNexus().Url)
 			if err != nil {
 				return nil, fmt.Errorf("failed to parse URL: %v: %w", &c, err)
 			}
-			baseURL = u.Scheme + "://" + u.Host
+			destination = u.Scheme + "://" + u.Host
+		case *persistencespb.Callback_Hsm:
```
Comment (bergundy) [inline components/callbacks/statemachine.go:107]: This is a bug fix

File: components/callbacks/executors.go [hunk 4]
```diff
@@ -160,9 +160,6 @@ func (e taskExecutor) loadInvocationArgs(
 				return err
 			}
 			invokable = hsmInvokable
-			if err != nil {
```
Comment (bergundy) [inline components/callbacks/executors.go:163]: This was redundant

File: components/nexusoperations/tasks.go [hunk 5]
```diff
@@ -48,8 +48,14 @@ func (TimeoutTask) Type() string {
 	return TaskTypeTimeout
 }
 
-func (t TimeoutTask) Kind() hsm.TaskKind {
-	return hsm.TaskKindTimer{Deadline: t.Deadline}
+// Deadline implements hsm.Task.
```
Comment (bergundy) [inline components/nexusoperations/tasks.go:51]:
```suggestion
```

File: components/nexusoperations/tasks.go [hunk 6]
```diff
@@ -48,8 +48,14 @@ func (TimeoutTask) Type() string {
...
+func (t TimeoutTask) Deadline() time.Time {
+	return t.deadline
+}
+
+// Destination implements hsm.Task.
```
Comment (bergundy) [inline components/nexusoperations/tasks.go:56]:
```suggestion
```

File: service/history/hsm/tasks.go [hunk 7]
```diff
@@ -49,8 +49,15 @@ var ErrInvalidTaskKind = errors.New("invalid task kind")
 type Task interface {
 	// Task type that must be unique per task definition.
 	Type() string
-	// Kind of the task, see [TaskKind] for more info.
-	Kind() TaskKind
+	// A deadline for firing this task.
+	// This represents a lower bound and actual execution may get delayed if the system is overloaded or for various
+	// other reasons.
+	// Return [NoDeadline] to schedule this task immediately.
```
Comment (bergundy) [inline service/history/hsm/tasks.go:55]:
```suggestion
	// Return [Immediate] to schedule this task immediately.
```

File: service/history/hsm/tasks.go [hunk 8]
```diff
@@ -49,8 +49,15 @@ var ErrInvalidTaskKind = errors.New("invalid task kind")
...
+	Deadline() time.Time
+	// The destination of this task, used to group tasks into a per namespace-and-destination scheduler.
+	// Tasks may leave
```
Comment (bergundy) [inline service/history/hsm/tasks.go:58]:
```suggestion
	// Tasks may return an empty string if the task is delivered internally within the system.
	// If a destination is set, a task will be scheduled on the outbound queue.
	// Currently Destination and Deadline are mutually exclusive.
```

**Review by yycptt — APPROVED** [review]
LGTM. We can talk later re. if component should/need the capability to pick a queue for its tasks.

**Pull request merged**